Fuzzer aot failure

```
module var1;
require UnitTest
bitfield var1{,}

[export]
def main{
 delete  makeSampleI ();
}
```

```
$ D=~/daScript2/daScript5 && clang++ -std=c++17 -fno-rtti -fno-exceptions -DNDEBUG -DDAS_NO_ASSERTIONS -DDAS_FUSION=2 -Wno-everything -fsyntax-only -I $D/include -I $D/3rdparty/fmt/include -I $D/modules/dasUnitTest t.cpp
t.cpp:87:5: error: no matching function for call to 'finalize_d2da0760e99565ae'
   87 |     finalize_d2da0760e99565ae(__context__,das_alias<SampleVariant>::to(das_arg<TVariant<16,8,int32_t,float,char *>>::pass((_temp_make_local_7_9_0 = (das_alias<SampleVariant>::from(makeSampleI()))))));
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~
t.cpp:79:13: note: candidate function not viable: no known conversion from 'SampleVariant' to 'TVariant<16, 8, int32_t, float, char *> &' (aka 'TVariant<16, 8, int, float, char *> &') for 2nd argument
   79 | inline void finalize_d2da0760e99565ae ( Context * __context__, TVariant<16,8,int32_t,float,char *> &  ____this_rename_at_7_0 )
      |             ^                                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.

```

# daslang AOT codegen bugs (found via grammar-fuzzing das_fuzz)

daslang version: `0.6.3` (commit 26420f77c). libclang: clang-19.

Each bug: `daslang -aot` succeeds (exit 0) but the **generated C++ fails to compile**. Repro is two commands — generate AOT C++, then compile it. `$D` = daslang root.

```bash
D=~/daScript2/daScript5   # daslang install root
```

## Bug A — AOT emits a method with no return type

Method declared inside a `struct/class template` is AOT-emitted as `inline DAS_COMMENT(auto) _Func...` — the return type macro expands to nothing, so the function has no return type. Needs export_all (`-aot-macros`) to emit uninstantiated template members.

**Reproduce:**
```bash
$D/bin/daslang -aot bug_A.das /tmp/out.cpp -aot-macros -dasroot $D
clang++ -std=c++17 -fno-rtti -fno-exceptions -DNDEBUG -DDAS_NO_ASSERTIONS -DDAS_FUSION=2 -Wno-everything -fsyntax-only \
  -I $D/include -I $D/3rdparty/fmt/include -I $D/modules/dasUnitTest /tmp/out.cpp
```

**`bug_A.das`:**
```
 struct  template var5{
 def var5{
}
}
def main{
}
```

**clang error:**
```
/tmp/r.cpp:64:26: error: a type specifier is required for all declarations
   64 | inline DAS_COMMENT(auto) _Funcvar5Tickvar5 ( Context * __context__, struct var5 & __self_rename_at_2_1 );
      | ~~~~~~                   ^
```

## Bug B — AOT emits malformed C++ (`expected expression`)

Generated C++ is syntactically invalid.

**Reproduce:**
```bash
$D/bin/daslang -aot bug_B.das /tmp/out.cpp -dasroot $D
clang++ -std=c++17 -fno-rtti -fno-exceptions -DNDEBUG -DDAS_NO_ASSERTIONS -DDAS_FUSION=2 -Wno-everything -fsyntax-only \
  -I $D/include -I $D/3rdparty/fmt/include -I $D/modules/dasUnitTest /tmp/out.cpp
```

**`bug_B.das`:**
```
[export]
def var3(var4:var5&){
}
 variant var5{
}
def main{
 unsafe {
}
}
```

**clang error:**
```
/tmp/r.cpp:62:73: error: expected expression
```

## Bug C — AOT `delete` of variant-alias: finalize() signature mismatch

`delete` of a value whose type is a variant alias emits a `finalize_*` call whose argument type (the alias `SampleVariant`) doesn't match the generated finalizer signature (`TVariant<...>&`).

**Reproduce:**
```bash
$D/bin/daslang -aot bug_C.das /tmp/out.cpp -dasroot $D
clang++ -std=c++17 -fno-rtti -fno-exceptions -DNDEBUG -DDAS_NO_ASSERTIONS -DDAS_FUSION=2 -Wno-everything -fsyntax-only \
  -I $D/include -I $D/3rdparty/fmt/include -I $D/modules/dasUnitTest /tmp/out.cpp
```

**`bug_C.das`:**
```
require UnitTest;
[export]
def main{
 delete  makeSampleI ();
}
```

**clang error:**
```
/tmp/r.cpp:86:5: error: no matching function for call to 'finalize_d2da0760e99565ae'
   86 |     finalize_d2da0760e99565ae(__context__,das_alias<SampleVariant>::to(das_arg<TVariant<16,8,int32_t,float,char *>>::pass((_temp_make_local_4_9_0 = (das_alias<SampleVariant>::from(makeSampleI()))))));
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~
```

## Bug E — AOT references an undefined local type (`__varN_rename...`)

Generated C++ references a local rename type (`__varN_rename_at_*`) that is never declared. Needs export_all.

**Reproduce:**
```bash
$D/bin/daslang -aot bug_E.das /tmp/out.cpp -aot-macros -dasroot $D
clang++ -std=c++17 -fno-rtti -fno-exceptions -DNDEBUG -DDAS_NO_ASSERTIONS -DDAS_FUSION=2 -Wno-everything -fsyntax-only \
  -I $D/include -I $D/3rdparty/fmt/include -I $D/modules/dasUnitTest /tmp/out.cpp
```

**`bug_E.das`:**
```
 struct  template var7{
 def var8(var7 = -5;
var4:var4-&,var8,var3: Point3 ){
}
}
def main{
}
```

**clang error:**
```
/tmp/r.cpp:63:130: error: unknown type name '__var7_rename_at_2_1'
```

## Bug F — AOT struct size mismatch (static_assert / ABI)

The AOT-generated struct's `sizeof` doesn't match the DAS runtime layout — a static_assert fires. This is an ABI mismatch, not just a syntax error: a real AOT build could miscompile silently.

**Reproduce:**
```bash
$D/bin/daslang -aot bug_F.das /tmp/out.cpp -dasroot $D
clang++ -std=c++17 -fno-rtti -fno-exceptions -DNDEBUG -DDAS_NO_ASSERTIONS -DDAS_FUSION=2 -Wno-everything -fsyntax-only \
  -I $D/include -I $D/3rdparty/fmt/include -I $D/modules/dasUnitTest /tmp/out.cpp
```

**`bug_F.das`:**
```
 class var2{
var8: :: var5 const ;
}
 struct var5;
def main{
}
```

**clang error:**
```
/tmp/r.cpp:65:15: error: static assertion failed due to requirement 'sizeof(das::_anon_15648304262845527893::(anonymous namespace)::var2) == 16': structure size mismatch with DAS
   65 | static_assert(sizeof(var2)==16,"structure size mismatch with DAS");
      |               ^~~~~~~~~~~~~~~~
```

## Bug G — AOT leaks generic placeholder `numT` into generated C++

The generated C++ contains the literal generic type-parameter name `numT` (an uninstantiated placeholder) instead of a concrete type.

**Reproduce:**
```bash
$D/bin/daslang -aot bug_G.das /tmp/out.cpp -dasroot $D
clang++ -std=c++17 -fno-rtti -fno-exceptions -DNDEBUG -DDAS_NO_ASSERTIONS -DDAS_FUSION=2 -Wno-everything -fsyntax-only \
  -I $D/include -I $D/3rdparty/fmt/include -I $D/modules/dasUnitTest /tmp/out.cpp
```

**`bug_G.das`:**
```
require UnitTest;
 var {
}
[export]
def main{
 return  array  variant <var5: variant <var4: urange >;var1: Point3 >();
}
```

**clang error:**
```
/tmp/r.cpp:87:255: error: use of undeclared identifier 'numT'
   94 |     return /* <- */ das_auto_cast_move<TArray<TVariant<16,4,TVariant<12,4,urange>,float3>>>::cast(_FuncbuiltinTickto_array_moveTick6262727710146066149_49d9d75c7cb8ce01(__context__,das_alias<TT>::to(das_arg<TVariant<16,4,TVariant<12,4,urange>,float3>>::pass((([&]() -> TVariant<16,4,TVariant<12,4,urange>,float3>& {
      |                                                                                                                                                                                               ^
```

---
_Note: Bugs A (**note from me: A is reproducible with no aot-macros**) and E require `-aot-macros` (= `export_all`); their buggy functions are uninstantiated template/generic members only emitted when all functions are force-AOT'd. B, C, F, G reproduce with plain `-aot`._

--- a/daslib/aot_cpp.das
+++ b/daslib/aot_cpp.das
@@ -397,6 +397,8 @@ def describeCppTypeEx(typeDecl : TypeDeclPtr;
                 ._fold()) |> join(",")
             if (cfg.cross_platform) {
                 write(writer, "AutoVariant<{types}>");
+            } elif (types |> empty()) {     // empty variant: no alternatives, only the index tag
+                write(writer, "TVariant<{typeDecl.variantSize},{typeDecl.variantAlign}>")
             } else {
                 write(writer, "TVariant<{typeDecl.variantSize},{typeDecl.variantAlign},{types}>")
             }

--- a/daslib/aot_cpp.das
+++ b/daslib/aot_cpp.das
@@ -387,6 +387,8 @@ def describeCppTypeEx(typeDecl : TypeDeclPtr;
                 ._fold()) |> join(",")
             if (cfg.cross_platform) {
                 write(writer, "AutoTuple<{types}>")
+            } elif (types |> empty()) {
+                write(writer, "TTuple<{typeDecl.tupleSize}>")
             } else {
                 write(writer, "TTuple<{typeDecl.tupleSize},{types}>")
             }

--- a/daslib/aot_cpp.das
+++ b/daslib/aot_cpp.das
@@ -1471,7 +1471,7 @@ class public CppAot : AstVisitor {
         ss |> string_builder_clear();
         write(*ss, "\n");
         prog.getThisModule |> for_each_module_function($(var fn : FunctionPtr) {
-            if (!fn.flags.builtIn && !fn.flags.noAot) {
+            if (!fn.flags.builtIn && !fn.flags.noAot && !fn.moreFlags.isTemplate) {
                 write(*ss, "{describeCppFunc(fn,collector,cross_platform)};\n");
             }
         });

--- a/daslib/aot_cpp.das
+++ b/daslib/aot_cpp.das
@@ -3756,7 +3756,8 @@ class public CppAot : AstVisitor {
         write(*ss, "{get_aot_arg_prefix(call.func, call, argIndex)}");
         assume argType = call.arguments[argIndex]._type;
         assume funArgType = call.func.arguments[argIndex]._type;
-        if (funArgType.isAotAlias) {
+        // alias wrap (structural -> alias spelling) is only correct for builtIn callees; emitter-generated callees spell params structurally
+        if (call.func.flags.builtIn && funArgType.isAotAlias) {
             if (funArgType.alias |> empty()) {
                 let fun_t_str = describeCppTypeEx(funArgType, DescribeConfig(skip_ref = true, redundant_const = true, cross_platform = cross_platform), CpptUseAlias.yes);
                 write(*ss, "das_reinterpret<{fun_t_str}>::pass(");
@@ -3817,7 +3818,7 @@ class public CppAot : AstVisitor {
         if (!call.func.flags.noPointerCast && needPtrCast(funArgType, arg._type, arg)) {
             write(*ss, ")");
         }
-        if (funArgType.isAotAlias) write(*ss, ")");
+        if (call.func.flags.builtIn && funArgType.isAotAlias) write(*ss, ")");
         write(*ss, "{get_aot_arg_suffix(call.func, call, argIndex)}");
         if (!last) {
             write(*ss, ",");

--- a/include/daScript/simulate/aot.h
+++ b/include/daScript/simulate/aot.h
@@ -1260,7 +1260,7 @@ namespace das {
         constexpr auto align = variant_align<TA...>();
         constexpr auto ihdr = round_up(TypeSize<int>::size, align);
         int sizes[] = { TypeSize<TA>::size..., 0 };
-        int ma = 0;
+        int ma = ihdr;  // a variant always carries its index tag, even with no alternatives
         for (size_t i = 0; i != sizeof...(TA); ++i)
             ma = std::max(ihdr + round_up(sizes[i], align), ma);
         return ma;
@@ -1364,6 +1364,23 @@ namespace das {
             memcpy ( (char *)this, &arr, variantSize );
         }
         TData data;
+    };
+
+    // empty variant: no alternatives, just the inherited 4-byte index tag (no payload).
+    // a dedicated specialization avoids the generic template's `char data[variantSize-4]`,
+    // which would be a zero-length array (non-standard) when variantSize == sizeof(int32_t).
+    template <int variantSize, int variantAlign>
+    struct alignas(variantAlign) TVariant<variantSize, variantAlign> : Variant {
+        static_assert(variantSize == variant_sizeof<>());
+        static_assert(variantAlign == variant_align<>());
+        TVariant() {}
+        TVariant(const TVariant & arr) { moveT(arr); }
+        TVariant(TVariant && arr) { moveT(arr); }
+        TVariant & operator = ( const TVariant & arr ) { moveT(arr); return *this; }
+        TVariant & operator = ( TVariant && arr ) { moveT(arr); return *this; }
+        __forceinline void moveT ( const TVariant & arr ) {
+            memcpy ( (char *)this, &arr, variantSize );
+        }
     };
 
     template <typename ...TA>

--- a/include/daScript/simulate/aot.h
+++ b/include/daScript/simulate/aot.h
@@ -1197,7 +1197,7 @@ namespace das {
     };
 
     template <typename... TA>
-    constexpr int max_alignof() { return std::max({TypeAlign<TA>::align...}); }
+    constexpr int max_alignof() { return std::max({TypeAlign<TA>::align..., 1}); }
 
     template <typename... TA>
     constexpr int variant_align() {
@@ -1291,6 +1291,16 @@ namespace das {
             memcpy ( data, &arr, tupleSize );
         }
         alignas(max_alignof<TA...>()) char data[tupleSize];
+    };
+
+    template <int tupleSize>
+    struct TTuple<tupleSize> : Tuple {
+        static_assert(tupleSize == 0);
+        TTuple() {}
+        TTuple(const TTuple &) {}
+        TTuple(TTuple &&) {}
+        TTuple & operator = ( const TTuple & ) { return *this; }
+        TTuple & operator = ( TTuple && ) { return *this; }
     };
 
     template <typename ...TA>

--- a/src/ast/ast.cpp
+++ b/src/ast/ast.cpp
@@ -446,7 +446,8 @@ namespace das {
                 }
             }
             size = (size + al) & ~al;
-            size += fd.type->getSizeOf64();
+            // an empty struct field is 0 bytes in daScript but occupies >=1 byte in C++.
+            size += das::max(uint64_t(1), fd.type->getSizeOf64());
         }
         circularGuard = false;
         int al = getAlignOf() - 1;
@@ -472,7 +473,7 @@ namespace das {
                 }
             }
             size = (size + al) & ~al;
-            size += fd.type->getSizeOf64(failed);
+            size += das::max(uint64_t(1), fd.type->getSizeOf64(failed));
         }
         circularGuard = false;
         int al = getAlignOfFailed(failed) - 1;

--- a/src/ast/ast_infer_type.cpp
+++ b/src/ast/ast_infer_type.cpp
@@ -424,7 +424,8 @@ namespace das {
             }
             fieldOffset = (fieldOffset + fa) & ~fa;
             decl.offset = int(fieldOffset);
-            fieldOffset += decl.type->getSizeOf64();
+            // An empty-struct field is 0 bytes here but >=1 in the AOT-emitted C++.
+            fieldOffset += das::max(uint64_t(1), decl.type->getSizeOf64());
         }
         verifyType(decl.type, false, decl.classMethod);
     }

--- a/src/ast/ast_typedecl.cpp
+++ b/src/ast/ast_typedecl.cpp
@@ -3035,7 +3035,8 @@ namespace das
 
     uint64_t TypeDecl::getVariantSize64() const {
         DAS_ASSERT(baseType==Type::tVariant);
-        uint64_t maxSize = 0;
+        // a variant always carries its index tag, even with no alternatives (empty variant)
+        uint64_t maxSize = getTypeBaseSize(Type::tInt);
         int al = getVariantAlign() - 1;
         for ( const auto & argT : argTypes ) {
             uint64_t size = (getTypeBaseSize(Type::tInt) + al) & ~al;
@@ -3048,7 +3049,7 @@ namespace das
 
     uint64_t TypeDecl::getVariantSize64(bool & failed) const {
         DAS_ASSERT(baseType==Type::tVariant);
-        uint64_t maxSize = 0;
+        uint64_t maxSize = getTypeBaseSize(Type::tInt);
         int al = getVariantAlignFailed(failed) - 1;
         for ( const auto & argT : argTypes ) {
             uint64_t size = (getTypeBaseSize(Type::tInt) + al) & ~al;

--- a/src/builtin/module_builtin_ast_serialize.cpp
+++ b/src/builtin/module_builtin_ast_serialize.cpp
@@ -1242,8 +1242,7 @@ namespace das {
             case tTuple:
             case tVariant:
                 ser << alias << argTypes << argNames;
-                DAS_VERIFYF_MULTI(!annotation, !structType, !enumType, !firstType, !secondType,
-                                !argTypes.empty());
+                DAS_VERIFYF_MULTI(!annotation, !structType, !enumType, !firstType, !secondType);
                 break;
             default:
                 SERIALIZER_VERIFYF(false,  "not expected to be here");

--- a/tests/README.md
+++ b/tests/README.md
@@ -39,6 +39,7 @@ Every `.das` file in this directory tree is listed below, grouped by subdirector
 | test_struct_decl_order.das | AOT required-module struct topo-sort (#3212) — by-value field of a later-declared struct | |
 | _aot_so_mod.das | *(helper)* #3212 fixture module — Outer holds Inner by value, declared Outer-first | |
 | t_invoke_void.das | AOT `invoke` dispatch — void-returning function pointer via AOT path | |
+| test_variant_alias_aot.das | AOT variant-alias codegen (#3269) — delete of handled variant alias, builtin push of a variant element | |
 
 ## apply/
 

--- a/tests/aot/test_empty_struct_field_aot.das
+++ b/tests/aot/test_empty_struct_field_aot.das
@@ -1,0 +1,24 @@
+options gen2
+require dastest/testing_boost public
+
+// Regression #3269 (bug F): an empty struct is 0 bytes standalone but occupies >=1 byte in C++
+// (no zero-size member). Embedding one by value used to make the AOT-emitted sizeof/offsetof
+// static_asserts diverge from the interpreter layout. Structure size/offset now count an
+// empty-struct field as >=1, so the emitted C++ matches. The empty field is placed first to
+// exercise the offsets of the fields that follow it.
+
+struct Marker {}
+
+struct Holder {
+    m : Marker
+    x : int
+    y : int
+}
+
+[test]
+def test_empty_struct_field_aot(t : T?) {
+    t |> equal(typeinfo sizeof(type<Marker>), 0)
+    let h = Holder(x = 7, y = 9)
+    t |> equal(h.x, 7)
+    t |> equal(h.y, 9)
+}

--- a/tests/aot/test_empty_tuple_aot.das
+++ b/tests/aot/test_empty_tuple_aot.das
@@ -1,0 +1,16 @@
+options gen2
+require dastest/testing_boost public
+
+tuple Empty {}
+
+struct Holder {
+    e : Empty
+    x : int
+}
+
+[test]
+def test_empty_tuple_aot(t : T?) {
+    t |> equal(typeinfo sizeof(type<Empty>), 0)
+    let h = Holder(x = 13)
+    t |> equal(h.x, 13)
+}

--- a/tests/aot/test_empty_variant_aot.das
+++ b/tests/aot/test_empty_variant_aot.das
@@ -1,0 +1,22 @@
+options gen2
+require dastest/testing_boost public
+
+// Regression #3269 (bug B): an empty variant has no alternatives, only its 4-byte index tag.
+// The AOT emitter used to write TVariant<0,4,> (trailing comma + size 0 -> char data[-4]). It now
+// emits TVariant<4,4> (the variant size counts the always-present index tag) and a dedicated
+// empty-variant C++ specialization compiles it without a zero-length data array. The empty variant
+// is also embedded by value to exercise the containing struct's layout.
+
+variant Empty {}
+
+struct Holder {
+    e : Empty
+    x : int
+}
+
+[test]
+def test_empty_variant_aot(t : T?) {
+    t |> equal(typeinfo sizeof(type<Empty>), 4)
+    let h = Holder(x = 13)
+    t |> equal(h.x, 13)
+}

--- a/tests/aot/test_variant_alias_aot.das
+++ b/tests/aot/test_variant_alias_aot.das
@@ -1,0 +1,43 @@
+options gen2
+require UnitTest
+require dastest/testing_boost public
+
+// Regression for #3269 (fuzzer AOT codegen). The AOT C++ emitter wrapped a call argument in
+// das_alias<Alias>::to(...) whenever the callee parameter was an alias type. That is only correct
+// for extern/bound (builtIn) callees whose C++ signature spells the alias name; emitter-generated
+// callees always spell parameters structurally, so the wrap mismatched the parameter and the
+// generated C++ failed to compile. bug C: delete of a handled variant alias (SampleVariant) ->
+// generated finalize(). bug G: builtin push/to_array_move of a variant element -> generic
+// placeholder name leaked. Both go through the now builtIn-gated alias wrap in daslib/aot_cpp.das.
+
+variant U_F {
+    i_value : uint
+    f_value : float
+}
+
+[sideeffects]
+def test_delete_handled_variant_alias(t : T?) {
+    var s = makeSampleS()
+    t |> success(s is s_value)
+    delete s
+    t |> success(s is i_value)
+}
+
+def test_variant_element_array(t : T?) {
+    var v : U_F
+    unsafe {
+        set_variant_index(v, 0)
+        v.i_value = 7u
+    }
+    var arr : array<U_F>
+    arr |> push(v)
+    t |> equal(length(arr), 1)
+    t |> success(arr[0] is i_value)
+    t |> equal(arr[0] as i_value, 7u)
+}
+
+[test]
+def test_variant_alias_aot(t : T?) {
+    test_delete_handled_variant_alias(t)
+    test_variant_element_array(t)
+}

